specification/securityandcompliance/resource-manager/Microsoft.SecurityAndCompliance/stable/2021-01-11/privateLinkServicesForEDMUpload.json

```diff
@@ -0,0 +1,1255 @@
+        "tags": [
+          "PrivateLinkResources"
+        ],
+        "operationId": "PrivateLinkResources_ListByService",
+        "description": "Gets the private link resources that need to be created for a service.",
+        "parameters": [
+          {
+            "$ref": "#/parameters/subscriptionId"
+          },
+          {
+            "$ref": "#/parameters/resourceGroupName"
+          },
+          {
+            "$ref": "#/parameters/resourceName"
+          },
+          {
+            "$ref": "#/parameters/ApiVersionParameter"
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "The request was successful; the request was well-formed and received properly.",
+            "schema": {
+              "$ref": "./privatelinks.json#/definitions/PrivateLinkResourceListResult"
+            }
+          },
+          "default": {
+            "description": "DefaultErrorResponse",
+            "schema": {
+              "$ref": "#/definitions/ErrorDetails"
+            }
+          }
+        },
+        "x-ms-examples": {
+          "PrivateLinkResources_ListGroupIds": {
+            "$ref": "./examples/EdmUploadPrivateLinkResourcesListByService.json"
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.SecurityAndCompliance/privateLinkServicesForEDMUpload/{resourceName}/privateLinkResources/{groupName}": {
+      "get": {
+        "tags": [
+          "PrivateLinkResources"
+        ],
+        "operationId": "PrivateLinkResources_Get",
+        "description": "Gets a private link resource that need to be created for a service.",
+        "parameters": [
+          {
+            "$ref": "#/parameters/subscriptionId"
+          },
+          {
+            "$ref": "#/parameters/resourceGroupName"
+          },
+          {
+            "$ref": "#/parameters/resourceName"
+          },
+          {
+            "$ref": "#/parameters/ApiVersionParameter"
+          },
+          {
+            "name": "groupName",
+            "in": "path",
+            "description": "The name of the private link resource group.",
+            "required": true,
+            "type": "string"
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "OK. The request was successful; the request was well-formed and received properly.",
+            "schema": {
+              "$ref": "./privatelinks.json#/definitions/PrivateLinkResource"
+            }
+          },
+          "default": {
+            "description": "DefaultErrorResponse",
+            "schema": {
+              "$ref": "#/definitions/ErrorDetails"
+            }
+          }
+        },
+        "x-ms-examples": {
+          "PrivateLinkResources_Get": {
+            "$ref": "./examples/EdmUploadPrivateLinkResourceGet.json"
+          }
+        }
+      }
+    }
+  },
+  "definitions": {
+    "ServicesProperties": {
```

Comment:
>ServicesProperties [](start = 5, length = 18)

there seems to be a lot of duplication of this schema in each of the files, could you not use the discriminator pattern? and then you don't need different resource types for each partner eg - EDM, M365, )3654 etc

Reply:
could you please share an example of discriminator pattern?

Also, as per our discussions with PM team, I think we'll need different resource types for each service.

Reply:
[https://swagger.io/specification/v2/](https://swagger.io/specification/v2/) search for discriminator, you can also search for discriminator and x-ms-discriminator-value in this repo

---
In reply to: [574272782](https://github.com/Azure/azure-rest-api-specs/pull/12812#discussion_r574272782) [](ancestors = 574272782)

Reply:
I don't see how different resource type names is a PM requirement - as long as the scenarios can be satisfied by the apis

---
In reply to: [574926775](https://github.com/Azure/azure-rest-api-specs/pull/12812#discussion_r574926775) [](ancestors = 574926775,574272782)

Reply:
So, we are creating private linkable resources (meaning against these resources private endpoints can be created). And we have multiple services against which we want to enable the creation of private endpoints. Do you recommend a discriminator pattern in this case?

These resource types will be shown to customers in the private endpoint creation wizard in the Azure portal. Users will be selecting one of these to create private endpoints.

Reply:
From what I can tell, Varun's suggestion satisfies your needs with a less complex API. The list you show the customer is the possible values the discriminator (e.g. kind-- probably an enum) can take, one for each of the multiple services you support. Your API specification would declare a separate schema for each of those, but the common properties would be declared once in the base schema. This implements your requirements using a single ARM resource type and a single corresponding endpoint path.

Reply:
Since, we need to show all these resource types in the dropdown in create private endpoint UI, I think we'll need all the resource types. Internally, we have implemented only one set of apis and the same endpoint URL will be configured for all these resource types in the manifest.